internal/compile/constraints.go

```diff
@@ -0,0 +1,226 @@
+	}
+
+	return &c, nil
+}
+
+// Builtin returns true if the builtin is supported by the constraint.
+func (c *Constraint) Builtin(x string) bool {
+	return c.Builtins.Contains(x)
+}
+
+func (c *Constraint) AssertBuiltin(x string) error {
+	if !c.Builtin(x) {
+		return fmt.Errorf("unsupported for %s", c)
+	}
+	return nil
+}
+
+// Supports allows us to encode more fluent constraints, like support for "not".
+// It returns an error if the feature is not supported by the constraint.
+func (c *Constraint) Supports(x string) bool {
+	return c.Features.Contains(x)
+}
+
+func (c *Constraint) AssertFeature(x string) error {
+	if !c.Supports(x) {
+		return fmt.Errorf("unsupported feature %q for %s", x, c)
+	}
+	return nil
+}
+
+var _ fmt.Stringer = (*Constraint)(nil)
+
+func (c *Constraint) String() string {
+	if c.Variant == "" {
+		return c.Target
+	}
+	return fmt.Sprintf("%s (%s)", c.Target, c.Variant)
+}
+
+type ConstraintSet struct {
+	Constraints []*Constraint
+}
+
+func NewConstraintSet(cs ...*Constraint) *ConstraintSet {
+	return &ConstraintSet{Constraints: cs}
+}
+
+// Builtin returns true if all the constraints in the set support the builtin.
+func (cs *ConstraintSet) Builtin(x string) bool {
+	for i := range cs.Constraints {
+		if !cs.Constraints[i].Builtin(x) {
+			return false
+		}
+	}
+	return true
+}
+
+func (cs *ConstraintSet) AssertBuiltin(x string) error {
+	var err error
+	for i := range cs.Constraints {
+		if e := cs.Constraints[i].AssertBuiltin(x); e != nil {
+			err = errors.Join(err, e)
+		}
+	}
+	return err
+}
+
+// Supports returns true if the feature is supported by all constraints.
+func (cs *ConstraintSet) Supports(x string) bool {
+	for i := range cs.Constraints {
+		if !cs.Constraints[i].Supports(x) {
+			return false
+		}
+	}
+	return true
+}
+
+// AssertFeature returns an error if the feature is not supported by any
+// of the constraints. The error contains the offending target/dialect pairs.
+func (cs *ConstraintSet) AssertFeature(x string) error {
+	var err error
+	for i := range cs.Constraints {
+		if e := cs.Constraints[i].AssertFeature(x); e != nil {
+			err = errors.Join(err, e)
+		}
+	}
+	return err
+}
+
+var _ fmt.Stringer = (*ConstraintSet)(nil)
+
+func (cs *ConstraintSet) String() string {
+	result := make([]string, 0, len(cs.Constraints)+1)
+	for i := range cs.Constraints {
+		result = append(result, cs.Constraints[i].String())
+	}
+	return "multi-constraint: " + strings.Join(result, ", ")
+}
+
+var (
+	sqlBuiltins = allBuiltins
+
+	// sqlite doesn't support startswith/endswith/contains
+	sqlSQLiteBuiltins = ucastBuiltins.Clone().Add("internal.member_2")
+
+	ucastBuiltins = NewSet(
+		"eq",
+		"neq",
+		"lt",
+		"lte",
+		"gt",
+		"gte",
+	)
+	allBuiltins = ucastBuiltins.Clone().Add(
+		"internal.member_2",
+		// "nin", // TODO: deal with NOT IN
+		"startswith",
+		"endswith",
+		"contains",
+	)
+	ucastLINQBuiltins = ucastBuiltins.Clone().Add(
+		"internal.member_2",
+		// "nin", // TODO: deal with NOT IN
+	)
+)
```

Comment:
These are the built-in functions support by the different targets. It's a prime spot for extensions and new cool features, expanding the fragment of rego policies we can translate into SQL etc.